docs/references/http-gateway-protocol-spec.md

```diff
@@ -0,0 +1,471 @@
+1. Case-insensitive search for the `IC-CertificationExpression` header.
+   - If no such header is found, verification fails.
+   - If the header value is not structured as per [the certificate expression header](#the-certificate-expression-header), verification fails.
+1. Perform [certificate validation](#certificate-validation).
+1. Parse the `version` field from the `IC-Certificate` header value as per [the certificate header](#the-certificate-header).
+   - If the `version` field is missing or equal to `1` then proceed with [legacy response verification](#legacy-response-verification).
+   - If the `version` field is equal to `2` then continue.
+   - Otherwise, verification fails.
+1. Validate that `expr_path` is the most specific expression path in the tree for the given request URL path.
+   - If the absence of a more specific path cannot be proven (a lookup of more specific paths must return `Absent`) as per [lookup](https://internetcomputer.org/docs/current/references/ic-interface-spec/#lookup), verification fails.
+1. Let `expr_hash` be the label of the node in the tree at path `expr_path`.
+   - If no such label exists, verification fails.
+   - If this node has sibling nodes, verification fails.
+   - If `expr_hash` does not match the sha256 hash of the `IC-CertificateExpression` header value, verification fails.
+   - If `no_certification` is set, verification succeeds.
+   - Let `response_hash` be the response hash calculated according to [Response Hash Calculation](#response-hash-calculation)
+   - If `no_request_certification` is set
+     - If the `expr_hash` label node has an empty leaf node at the subpath `[response_hash]`, verification succeeds.
+     - Otherwise, the certification is invalid.
+   - Let `request_hash` be the request hash calculated according to [Request Hash Calculation](#request-hash-calculation)
+     - If there is not an empty leaf node at the subpath `[request_hash, response_hash]`, verification fails.
+
+### The Certificate Header
+
+The `IC-Certificate` header is a structured header according to [RFC 8941](https://www.rfc-editor.org/rfc/rfc8941.html) with the following mandatory fields:
+
+- `certificate`: [Base64 encoded](https://www.rfc-editor.org/rfc/rfc4648#section-4) string of self-describing, [CBOR-encoded](https://www.rfc-editor.org/rfc/rfc8949.html) bytes that decodes into a valid [certificate](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certification).
+- `tree`: [Base64 encoded](https://www.rfc-editor.org/rfc/rfc4648#section-4) string of self-describing, [CBOR-encoded](https://www.rfc-editor.org/rfc/rfc8949.html) bytes that decodes into a valid hash tree as per [certificate encoding](https://internetcomputer.org/docs/current/references/ic-interface-spec/#certification-encoding).
+
+The following additional fields are mandatory for response verification version 2 and upwards:
+
+- `version`
+- `expr_path`
```

Comment:
Maybe we can add some info for the type: e.g., integer indicating the version, string containing the path or something like that.

Reply:
I agree this is missing, I'll add it, thanks